prediction_market_agent/agents/goal_manager.py

```diff
@@ -0,0 +1,307 @@
+            f"Reasoning: {self.reasoning}\n"
+            f"Output: {self.output}"
+        )
+
+    @classmethod
+    def from_model(cls, model: EvaluatedGoalModel) -> "EvaluatedGoal":
+        return EvaluatedGoal(
+            goal=model.goal,
+            motivation=model.motivation,
+            completion_criteria=model.completion_criteria,
+            is_complete=model.is_complete,
+            reasoning=model.reasoning,
+            output=model.output,
+        )
+
+    def to_model(self, agent_id: str) -> EvaluatedGoalModel:
+        return EvaluatedGoalModel(
+            goal=self.goal,
+            motivation=self.motivation,
+            completion_criteria=self.completion_criteria,
+            is_complete=self.is_complete,
+            reasoning=self.reasoning,
+            output=self.output,
+            agent_id=agent_id,
+            datetime_=utcnow(),
+        )
+
+    def to_goal(self) -> Goal:
+        return Goal(
+            goal=self.goal,
+            motivation=self.motivation,
+            completion_criteria=self.completion_criteria,
+        )
+
+
+class GoalManager:
+    def __init__(
+        self,
+        agent_id: str,
+        high_level_description: str,
+        agent_capabilities: str,
+        retry_limit: int = 3,
+        model: str = DEFAULT_OPENAI_MODEL,
+        sqlalchemy_db_url: str | None = None,
+    ):
+        self.agent_id = agent_id
+        self.high_level_description = high_level_description
+        self.agent_capabilities = agent_capabilities
+        self.retry_limit = retry_limit
+        self.model = model
+        self.table_handler = EvaluatedGoalTableHandler(
+            agent_id=agent_id,
+            sqlalchemy_db_url=sqlalchemy_db_url,
+        )
+
+    def get_latest_evaluated_goals_from_memory(self, limit: int) -> list[EvaluatedGoal]:
+        evaluated_goal_models = self.table_handler.get_latest_evaluated_goals(
+            limit=limit
+        )
+        return [EvaluatedGoal.from_model(model) for model in evaluated_goal_models]
+
+    @observe()
+    def generate_goal(self, latest_evaluated_goals: list[EvaluatedGoal]) -> Goal:
+        """
+        Generate a new goal based on the high-level description and the latest
+        evaluated goals.
+
+        TODO support generation of long-horizon goals with a specified
+        completion date, until which the goal's status is 'pending'.
+        """
+        parser = PydanticOutputParser(pydantic_object=Goal)
+        prompt = PromptTemplate(
+            template=GENERATE_GOAL_PROMPT_TEMPLATE,
+            input_variables=[
+                "high_level_description",
+                "agent_capabilities",
+                "previous_evaluated_goals",
+            ],
+            partial_variables={"format_instructions": parser.get_format_instructions()},
+        )
+        latest_evaluated_goals_str = self.evaluated_goals_to_str(latest_evaluated_goals)
+        llm = ChatOpenAI(
+            temperature=0,
```

Comment:
This might be better with super low temperature constant I introduced somewhere, if you want consistency as much as possible

Reply:
I still don't feel totally convinced by the "very small temperature is best for consistency" argument (I guess the same point expressed in the comment thread on the original PR https://github.com/gnosis/prediction-market-agent-tooling/pull/307/files#r1675590362).

I re-googled to see if there was any new info, and haven't found anything to convince one way or the other. In the anthropic API docs (https://docs.anthropic.com/en/api/complete) it says:
```
Defaults to 1.0. Ranges from 0.0 to 1.0. Use temperature closer to 0.0 for analytical / multiple choice, and closer to 1.0 for creative and generative tasks.

Note that even with temperature of 0.0, the results will not be fully deterministic.
```
which suggests 0.0 is more consistent than 0.000001, but still too vague to be sure!

Anyway, all that to say, if it's alright with you I'll stick with 0.0 until we know more (and we just have inconsistent code 😅)

Reply:
Sounds good. I'm now curious to do some manual experiment about this 🙈